test/unit/org/apache/cassandra/db/marshal/AbstractTypeTest.java

```diff
@@ -740,4 +801,927 @@ public String toString()
+    }
+
+    public void testAssumedPrimitiveTypesCompatibility(TypesCompatibility typesCompatibility)
+    {
+        SoftAssertions assertions = new SoftAssertions();
+
+        // assumed compatibility
+        typesCompatibility.compatibleWith.forEach((l, r) -> assertions.assertThat(l.isCompatibleWith(r)).describedAs(isCompatibleWithDesc(l, r)).isTrue());
+        typesCompatibility.valueCompatibleWith.forEach((l, r) -> assertions.assertThat(l.isValueCompatibleWith(r)).describedAs(isValueCompatibleWithDesc(l, r)).isTrue());
+        typesCompatibility.serializationCompatibleWith.forEach((l, r) -> assertions.assertThat(l.isSerializationCompatibleWith(r)).describedAs(isSerializationCompatibleWithDesc(l, r)).isTrue());
+
+        // assumed incompatibility
+        inverse(typesCompatibility.compatibleWith).forEach(p -> assertions.assertThat(p.left.isCompatibleWith(p.right)).describedAs(isCompatibleWithDesc(p.left, p.right)).isFalse());
+        inverse(typesCompatibility.valueCompatibleWith).forEach(p -> assertions.assertThat(p.left.isValueCompatibleWith(p.right)).describedAs(isValueCompatibleWithDesc(p.left, p.right)).isFalse());
+        inverse(typesCompatibility.serializationCompatibleWith).forEach(p -> assertions.assertThat(p.left.isSerializationCompatibleWith(p.right)).describedAs(isSerializationCompatibleWithDesc(p.left, p.right)).isFalse());
+
+        // it is implied that isSerializationCompatibleWith is a subset of isValueCompatibleWith
+        typesCompatibility.serializationCompatibleWith.forEach((l, r) -> assertions.assertThat(typesCompatibility.valueCompatibleWith.containsEntry(l, r)).describedAs(isValueCompatibleWithDesc(l, r)).isTrue());
+        assertions.assertAll();
+    }
+
+    /**
+     * Assuming that {@link #testAssumedPrimitiveTypesCompatibility()} passes, this test case verifies whether the types
+     * said to be compatible hold the assumed properties. In particular:
+     * <li>L {@link AbstractType#isValueCompatibleWith(AbstractType)} R - if it is possible to {@code L.compose(R.decompose(v))} for any v valid for type R, and the converted value of type L still makes sense</li>
+     * <li>L {@link AbstractType#isSerializationCompatibleWith(AbstractType)} R - if it is possible to read a cell written using R's type serializer with L's type serializer</li>
+     * <li>L {@link AbstractType#isCompatibleWith(AbstractType)} R - L isSerializationCompatibleWith R and it is possible to compare decomposed values of R's type using L's type comparator</li>
+     */
+    @Test
+    public void testPrimitiveTypesCompatibility()
+    {
+        SoftAssertions assertions = new SoftAssertions();
+        typesCompatibility.valueCompatibleWith.forEach((left, right) -> assertions.check(() -> verifyTypesCompatibility(left, right, getTypeSupport(right).valueGen)));
+        assertions.assertAll();
+    }
+
+    private static <L, R> void verifyTypesCompatibility(AbstractType left, AbstractType right, Gen rightGen)
```

Comment:
```suggestion
    private static <L, R> void verifyTypesCompatibility(AbstractType<L> left, AbstractType<R> right, Gen<R> rightGen)
```

why have `<L, R>` if you use raw types?